bionic/cache.py

```diff
@@ -494,25 +494,47 @@ def _exact_metadata_url_for_query(self, query):
     def _load_metadata_from_url(self, url):
         try:
             metadata_yaml = self._fs.read_bytes(url).decode('utf8')
-            return ArtifactMetadataRecord.from_yaml(metadata_yaml)
+            record = ArtifactMetadataRecord.from_yaml(metadata_yaml)
+            record.artifact_url =\
+                Inventory._absolute_url(record.artifact_url, url)
+            return record
         except Exception as e:
             raise InternalCacheStateError.from_failure(
                 'metadata record', url, e)
 
     def _create_and_write_metadata(self, query, artifact_url):
+        metadata_url = self._exact_metadata_url_for_query(query)
+        artifact_url =\
+            Inventory._absolute_or_relative_url(artifact_url, metadata_url)
+
         metadata_record = ArtifactMetadataRecord.from_content(
             entity_name=query.entity_name,
             artifact_url=artifact_url,
             provenance=query.provenance,
         )
 
-        metadata_url = self._exact_metadata_url_for_query(query)
-
         self._fs.write_bytes(
             metadata_record.to_yaml().encode('utf8'), metadata_url)
 
         return metadata_url
 
+    # when both urls start with file://, make the artifact url relative to
+    # metadata url
+    @staticmethod
+    def _absolute_or_relative_url(artifact_url, metadata_url):
```

Comment:
nit: I wonder if these would be clearer as something like `relativize_url`/`derelativize_url`? Doesn't actually make their behavior more obvious but it does call out the relationship between them. Could go either way though.